erpnext/regional/italy/utils.py

```diff
@@ -0,0 +1,279 @@
+
+	for row in doc.items:
+		tax_rate = 0.0
+		if itemised_tax.get(row.item_code):
+			tax_rate = sum([tax.get('tax_rate', 0) for d, tax in itemised_tax.get(row.item_code).items()])
+
+		row.tax_rate = flt(tax_rate, row.precision("tax_rate"))
+		row.tax_amount = flt((row.net_amount * tax_rate) / 100, row.precision("net_amount"))
+		row.total_amount = flt((row.net_amount + row.tax_amount), row.precision("total_amount"))
+
+@frappe.whitelist()
+def export_invoices(filters=None):
+	saved_xmls = []
+
+	invoices = frappe.get_all("Sales Invoice", filters=get_conditions(filters), fields=["*"])
+
+	for invoice in invoices:
+		attachments = get_e_invoice_attachments(invoice)
+		saved_xmls += [attachment.file_name for attachment in attachments]
+
+	zip_filename = "{0}-einvoices.zip".format(frappe.utils.get_datetime().strftime("%Y%m%d_%H%M%S"))
+
+	download_zip(saved_xmls, zip_filename)
+
+	cleanup_files(saved_xmls)
+
+@frappe.whitelist()
+def prepare_invoice(invoice, progressive_number):
+	#set company information
+	company = frappe.get_doc("Company", invoice.company)
+
+	invoice["progressive_number"] = progressive_number
+	invoice["unamended_name"] = get_unamended_name(invoice)
+	invoice["company_data"] = company
+	company_address = frappe.get_doc("Address", invoice.company_address)
+	invoice["company_address_data"] = company_address
+
+	#Set invoice type
+	if invoice.is_return and invoice.return_against:
+		invoice["type_of_document"] = "TD04" #Credit Note (Nota di Credito)
+		invoice["return_against_unamended"] =  get_unamended_name(frappe.get_doc("Sales Invoice", invoice.return_against))
+	else:
+		invoice["type_of_document"] = "TD01" #Sales Invoice (Fattura)
+
+	#set customer information
+	invoice["customer_data"] = frappe.get_doc("Customer", invoice.customer)
+	customer_address = frappe.get_doc("Address", invoice.customer_address)
+	invoice["customer_address_data"] = customer_address
+
+	if invoice.shipping_address_name:
+		invoice["shipping_address_data"] = frappe.get_doc("Address", invoice.shipping_address_name)
+
+	if invoice["customer_data"].is_public_administration:
+		invoice["transmission_format_code"] = "FPA12"
+	else:
+		invoice["transmission_format_code"] = "FPR12"
+
+
+	items = frappe.get_all("Sales Invoice Item", filters={"parent":invoice.name}, fields=["*"], order_by="idx")
+	taxes = frappe.get_all("Sales Taxes and Charges", filters={"parent":invoice.name}, fields=["*"], order_by="idx")
+	tax_data = get_invoice_summary(items, taxes)
+
+	invoice["tax_data"] = tax_data
+
+	#Check if stamp duty (Bollo) of 2 EUR exists.
+	stamp_duty_charge_row = next((tax for tax in taxes if tax.charge_type == _("Actual") and tax.tax_amount == 2.0 ), None)
+	if stamp_duty_charge_row:
+		invoice["stamp_duty"] = stamp_duty_charge_row.tax_amount
+
+	#append items, and tax exemption reason.
+	for item in items:
+		if item.tax_rate == 0.0:
+			item["tax_exemption_reason"] = tax_data["0.0"]["tax_exemption_reason"]
+
+	invoice["invoice_items"] = items
```

Comment:
In template you can use doc.items? If yes then no need of it, also for the payment_terms